pkg/bytesutil/bytesutil.go

```diff
@@ -4,8 +4,82 @@ import (
 	"bytes"
 	"fmt"
 	"sort"
+	"sync"
 )
 
+// CircularBuffer is an io.Writer backed by a circular buffer, whose size can be
+// specified during initialization.
+//
+// CircularBuffer is safe for use from multiple goroutines.
+type CircularBuffer struct {
+	mu  sync.RWMutex
+	buf []byte // Underlying data.
+	n   int    // CircularBuffer capacity.
+	end int    // Current length of data in buffer.
+}
+
+// NewCircularBuffer creates and initialises a new CircularBuffer of size n.
+func NewCircularBuffer(n int) *CircularBuffer {
+	return &CircularBuffer{
+		buf: make([]byte, n),
+		n:   n,
+	}
+}
+
+// Write appends the contents of p to the CircularBuffer, returning the number of bytes
+// written, and any error encountered.
+func (b *CircularBuffer) Write(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if len(p) >= b.n { // p is larger than maximum capacity.
+		copy(b.buf, p[len(p)-b.n:])
+		b.end = b.n
+		return b.n, nil
+	}
+
+	if len(p)+b.end <= b.n { // p fits into existing capacity.
+		copy(b.buf[b.end:], p)
+		b.end += len(p)
+		return len(p), nil
+	}
+
+	// Move buffer down such that `end` is `len(p)` further down the buffer.
+	remain := b.n - b.end
+	copy(b.buf, b.buf[len(p)-remain:])
```

Comment:
I'm concerned this is going to be expensive since once the buffer is full, we'll always be shifting ~2MB of memory around under a global write lock when something is logged.  Did you look at using a circular buffer with start and end indexes into `p.buf`.  See [Circular Buffer](https://en.wikipedia.org/wiki/Circular_buffer)

Reply:
@jwilder I benchmarked the worst case—writing to a full buffer—and it came in at `0.41 ms` on my laptop, which is worse than I expected. I'll re-implement it as a full ring (I was prototyping this for internal discussion to be honest anyway).